### PR TITLE
fix: 🐞 remove redundant xywh2xyxy conversion in OBBValidator.plot_predictions

### PR DESCRIPTION
**Pull request by onuralpszr**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Improves OBB (oriented bounding box) validation plotting by switching to a new batched plotting path and handling empty predictions safely 🧭🖼️

### 📊 Key Changes
- Updated `OBBValidator.plot_predictions()` to accept predictions as a list of dictionaries (e.g., `{"bboxes", "cls", "conf"}`) instead of raw tensors 🧩
- Added an early return when `preds` is empty to avoid errors during validation runs 🚫
- Replaced the previous `super().plot_predictions()` call (and the duplicated bbox conversion logic) with direct use of `plot_images()` from `ultralytics.utils.plotting` 🎨
- Added `batch_idx` to each prediction and concatenated per-image predictions into a single batched structure before plotting 📦

### 🎯 Purpose & Impact
- Produces more reliable validation preview images for OBB tasks, especially when some batches have no detections ✅
- Aligns OBB plotting with a consistent “batched labels/preds” format used by `plot_images()`, reducing format mismatches and edge-case crashes 🔧
- Simplifies and centralizes plotting logic, making it easier to maintain and extend in future updates 🛠️

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @onuralpszr, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This is an automated review assistant, and a Ultralytics engineer will be with you shortly. In the meantime, please review the checklist below to help ensure a smooth merge:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the **Update branch** button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing `I have read the CLA Document and I sign the CLA` in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If you have any questions or need clarification about OBB plotting behavior or validator changes, feel free to leave a comment. Thank you again for contributing to Ultralytics! ✨🚀

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Add an early return when no predictions are available so validation plotting doesn’t crash on empty batches.

💬 Posted 1 inline comment

**Comment by onuralpszr** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch for this PR at 2025-11-23 15:08:07 UTC with `@ultralytics/run-ci` command:

* ✅ [CI](https://github.com/ultralytics/ultralytics/actions/runs/19613075995): `ci.yml` run 14691
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/19613076133): `docker.yml` run 1443

File: ultralytics/models/yolo/obb/val.py [hunk 1]
```diff
@@ -141,24 +142,34 @@ def _prepare_batch(self, si: int, batch: dict[str, Any]) -> dict[str, Any]:
...
         """
-        for p in preds:
-            # TODO: fix this duplicated `xywh2xyxy`
-            p["bboxes"][:, :4] = ops.xywh2xyxy(p["bboxes"][:, :4])  # convert to xyxy format for plotting
-        super().plot_predictions(batch, preds, ni)  # plot bboxes
+        if not preds:
+            return
+        for i, pred in enumerate(preds):
+            pred["batch_idx"] = torch.ones_like(pred["conf"]) * i
+        keys = preds[0].keys()
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/obb/val.py:163]: ⚠️ **HIGH**: `torch.cat([x[k] for x in preds], dim=0)` assumes that every key in `preds[0]` is present in all other prediction dicts and contains a tensor. In practice prediction dicts often include optional entries such as `masks`, `probs`, or `boxes_3d` that may be `None` or have inconsistent shapes, so this line will raise at runtime when it encounters a non-tensor value. Please filter to tensor keys (and skip `None`) before concatenation or handle those fields individually to keep plotting from crashing.

**Suggested change:**
```suggestion
        batched_preds = {k: torch.cat([x[k] for x in preds if x.get(k) is not None], dim=0) for k in preds[0] if torch.is_tensor(preds[0][k])}
```

File: ultralytics/models/yolo/obb/val.py [hunk 2]
```diff
@@ -141,24 +142,34 @@ def _prepare_batch(self, si: int, batch: dict[str, Any]) -> dict[str, Any]:
...
             >>> validator.plot_predictions(batch, preds, 0)
         """
-        for p in preds:
-            # TODO: fix this duplicated `xywh2xyxy`
-            p["bboxes"][:, :4] = ops.xywh2xyxy(p["bboxes"][:, :4])  # convert to xyxy format for plotting
-        super().plot_predictions(batch, preds, ni)  # plot bboxes
+        if not preds:
+            return
+        for i, pred in enumerate(preds):
+            pred["batch_idx"] = torch.ones_like(pred["conf"]) * i
```
Comment (UltralyticsAssistant) [inline ultralytics/models/yolo/obb/val.py:162]: 💡 **MEDIUM**: `torch.ones_like(pred["conf"]) * i` creates a float tensor for `batch_idx`, but `plot_images` treats `batch_idx` as integer indices. Storing floats can lead to precision issues when comparing against integer batch selectors and is inconsistent with the rest of the pipeline. Use an integer tensor (e.g., `torch.full`) with `dtype=torch.long` to avoid subtle indexing bugs.

**Suggested change:**
```suggestion
            pred["batch_idx"] = torch.full_like(pred["cls"], i, dtype=torch.long)
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review 2

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Ensure batching skips non-tensor prediction fields to avoid crashes and store `batch_idx` as integers to keep downstream indexing stable.

💬 Posted 2 inline comments

**Review by glenn-jocher — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
Thrilled to see this PR merged—thank you @onuralpszr (and @glenn-jocher for the contributions) for the thoughtful upgrades to OBB validation plotting.

As Leonardo da Vinci said, *“Simplicity is the ultimate sophistication.”* This change embodies that: moving to the batched `plot_images()` path, safely handling empty predictions, and standardizing the `{bboxes, cls, conf}` dict format makes OBB validation previews more reliable and the plotting code far easier to maintain and extend.

Appreciate you both helping make YOLO11 OBB workflows sturdier for everyone.
